include/mcl_3dl/raycasts/raycast_using_dda.h

```diff
@@ -0,0 +1,292 @@
+  {
+  }
+
+  void setRay(typename ChunkedKdtree<POINT_TYPE>::Ptr kdtree, const Vec3 ray_begin, const Vec3 ray_end_org) final
+  {
+    kdtree_ = kdtree;
+    updatePointCloud();
+    ray_begin_ = ray_begin;
+    ray_direction_vector_ = (ray_end_org - ray_begin).normalized();
+    const Vec3 ray_end = ray_direction_vector_ * additional_range_ + ray_end_org;
+    begin_index_ = toIndex(ray_begin);
+    end_index_ = toIndex(ray_end);
+    const Eigen::Vector3i distance_index = end_index_ - begin_index_;
+    pos_ = 0;
+    max_movement_ = std::abs(distance_index[0]) + std::abs(distance_index[1]) + std::abs(distance_index[2]);
+    step_ = Eigen::Vector3i((distance_index[0] < 0) ? -1 : 1, (distance_index[1] < 0) ? -1 : 1,
+                            (distance_index[2] < 0) ? -1 : 1);
+
+    current_index_ = begin_index_;
+    for (int i = 0; i < 3; ++i)
+    {
+      if (distance_index[i] == 0)
+      {
+        initial_edges_[i] = std::numeric_limits<double>::infinity();
+        t_delta_[i] = std::numeric_limits<double>::infinity();
+      }
+      else
+      {
+        const double nearest = (ray_direction_vector_[i] < 0) ? begin_index_[i] * dda_grid_size_ + min_p_[i] :
+                                                                (begin_index_[i] + 1) * dda_grid_size_ + min_p_[i];
+        initial_edges_[i] = std::abs((nearest - ray_begin[i]) / ray_direction_vector_[i]);
+        t_delta_[i] = std::abs(dda_grid_size_ / ray_direction_vector_[i]);
+      }
+    }
+    t_max_ = initial_edges_;
+  }
+
+  bool getNextCastResult(CastResult& result) final
+  {
+    if (pos_ >= max_movement_)
+    {
+      return false;
+    }
+    ++pos_;
+
+    if (t_max_[0] < t_max_[1])
+    {
+      if (t_max_[0] < t_max_[2])
+      {
+        if (!incrementIndex(0))
+        {
+          return false;
+        }
+      }
+      else
+      {
+        if (!incrementIndex(2))
+        {
+          return false;
+        }
+      }
+    }
+    else
+    {
+      if (t_max_[1] < t_max_[2])
+      {
+        if (!incrementIndex(1))
+        {
+          return false;
+        }
+      }
+      else
+      {
+        if (!incrementIndex(2))
+        {
+          return false;
+        }
+      }
+    }
+    const POINT_TYPE* collided_point = hasIntersection(current_index_);
+    if (collided_point)
+    {
+      // TODO(nhatao): Implement estimation of the angle of incidence.
+      result = {fromIndex(current_index_), true, 1.0, collided_point};
+    }
+    else
+    {
+      result = {fromIndex(current_index_), false, 1.0, collided_point};
+    }
+    return true;
+  }
+
+private:
+  void updatePointCloud()
+  {
+    if (!kdtree_->getInputCloud())
+    {
+      return;
+    }
+    if ((kdtree_->getInputCloud()->header.stamp == point_cloud_header_.stamp) && !points_.empty())
+    {
+      return;
+    }
+    const typename pcl::PointCloud<POINT_TYPE>::ConstPtr& input = kdtree_->getInputCloud();
```

Comment:
Converting whole map into gridmap requires large memory.
e.g. size of a largest map I have is 468x308x29m. It requires 2,339 * 1,540 * 147 * (1 (point_exists_) + 8 (points_)) = 530M * 9 = 4.7GB for 0.2m grid.
(pcd file size is 28MB)

This was why I didn't do it in the Kd-tree based implementation.

Reply:
As an experimental feature, it's ok for now.

Reply:
The size of `points_` depends on the number of points of `pc_map_`(after downsampled), so additional memory consumption will become 530MB + maybe less than 10MB. 
To use this algorithm for larger map, it is needed to implement another method to reduce memory sizes.